zebra-scan/src/scan.rs

```diff
@@ -80,115 +81,159 @@ pub async fn start(
             Ok::<_, Report>((key.clone(), parsed_keys))
         })
         .try_collect()?;
+    let parsed_keys = Arc::new(parsed_keys);
 
     // Give empty states time to verify some blocks before we start scanning.
     tokio::time::sleep(INITIAL_WAIT).await;
 
     loop {
-        // Get a block from the state.
-        // We can't use ServiceExt::oneshot() here, because it causes lifetime errors in init().
-        let block = state
-            .ready()
-            .await
-            .map_err(|e| eyre!(e))?
-            .call(zebra_state::Request::Block(height.into()))
-            .await
-            .map_err(|e| eyre!(e))?;
-
-        let block = match block {
-            zebra_state::Response::Block(Some(block)) => block,
-            zebra_state::Response::Block(None) => {
-                // If we've reached the tip, sleep for a while then try and get the same block.
-                tokio::time::sleep(CHECK_INTERVAL).await;
-                continue;
-            }
-            _ => unreachable!("unmatched response to a state::Tip request"),
-        };
-
-        // Only log at info level every 100,000 blocks
-        let is_info_log =
-            height == storage.min_sapling_birthday_height() || height.0 % INFO_LOG_INTERVAL == 0;
-
-        // TODO: add debug logs?
+        let scanned_height = scan_height_and_store_results(
+            height,
+            state.clone(),
+            chain_tip_change.clone(),
+            storage.clone(),
+            key_birthdays.clone(),
+            parsed_keys.clone(),
+        )
+        .await?;
+
+        // If we've reached the tip, sleep for a while then try and get the same block.
+        if scanned_height.is_none() {
+            tokio::time::sleep(CHECK_INTERVAL).await;
+            continue;
+        }
+
+        height = height
+            .next()
+            .expect("a valid blockchain never reaches the max height");
+    }
+}
+
+/// Get the block at `height` from `state`, scan it with the keys in `parsed_keys`, and store the
+/// results in `storage`. If `height` is lower than the `key_birthdays` for that key, skip it.
+///
+/// Returns:
+/// - `Ok(Some(height))` if the height was scanned,
+/// - `Ok(None)` if the height was not in the database, and
```

Comment:
```suggestion
/// - `Ok(None)` if the height was not in the state, and
```

Reply:
Fair point, since we have multiple databases now.